### PR TITLE
Deprecate all Spatial object/sp functionality 

### PR DESCRIPTION
**Pull request by stephhazlitt**
Address part of #109.

- adds note(s) re: deprecation of Spatial object functionality to README
- adds warning on package load re: deprecation of Spatial object fxns (smelly?)
- uses `.Deprecate` in the `convert_to_sp()` helper fxn in `bcmaps::get_layer()`, which is triggered with _all_ the named layer + shortcut functions e.g. `bc_bound(class = "sp")`, `bc_neighbours(class = "sp")`
- uses `.Deprecate` in the methods helper fxns (`transform_bc_albers.Spatial` and`fix_geo_problems.Spatial`)
- uses `.Deprecate` in the `Spatial` helper fxns `get_poly_attribute` and `self_union`

The Approach:
 - [x] I am not sure about the `.onAttach` business, it was a shot in the dark 😬
 - [x] `combine_nr_rd()` returns the warning 2x because `convert_to_sp()` is [called on two layers](https://github.com/bcgov/bcmaps/blob/master/R/utils.R#L335-L336)
 - [x] I am not sure how to tackle `bc_bbox()` with `.Deprecate`, we could use [{lifecycle}](https://lifecycle.r-lib.org/reference/deprecate_soft.html)?
 - [x] This approach returns a warning for users for each/most `sp` functions, but does not alert users via the package documentation. We could add {lifecycle} as a dependency and leverage the lifecycle badges and Roxygen documentation (and the argument-level functionality for the above situation)&mdash;we would need to edit the [`make_shortcuts()`](https://github.com/bcgov/bcmaps/blob/master/R/make_shortcuts.R) fxn and then regenerate the `shortcuts.R` script if we want to go that route.

### CODE REVIEW

File: R/zzz.R [hunk 1]
```diff
@@ -16,3 +16,17 @@ bcmaps_env <- new.env(parent = emptyenv())
...
+bcmapsStartupMessage <- function() {
+  msg <- "Support for Spatial objects (`sp`) was deprecated in {bcmaps} v1.2.0. Please use `sf` objects with {bcmaps}."
+}
+
+.onAttach <- function(lib, pkg)
+{
+  # startup message
+  msg <- bcmapsStartupMessage()
+  if(!interactive())
+    msg[1]
```
Comment (ateucher) [inline R/zzz.R:29]: I don't think this is necessary?

**Review by ateucher — COMMENTED** [review]
I think this looks awesome! Great call on doing I do wonder about using lifecycle throughout (`deprecate_warn()`), as I think it gives us a few advantages:

1. It only warns users once every eight hours, I think with so many functions having this deprecation it could get annoying (and will avoid the annoying double message in `combine_nr_rd()`).
2. It gives us the ability to note the deprecation clearly in the function documentation (as you mention) - specifically the `class` argument. (https://lifecycle.r-lib.org/articles/communicate.html#arguments)
3. We can test it via `expect_deprecated()`: https://lifecycle.r-lib.org/reference/expect_deprecated.html
4. We can switch to `deprecate_stop()` easily down the road when we are ready to make it defunct.

As you say, this will require editing `make_shortcuts.R` and re-running it.

I think it's worth making a helper function (e.g., `deprecate_sp()`) that calls `lifecycle::deprecate_warn()` and use it throughout to avoid repetition - and replace all the uses of the `.Deprecated()` with it. This could have an argument that flags whether it is called from a `.Spatial` method or as a result of setting `class = "sp"` and have slightly different messages for each... or not if that is unnecessarily complex?

I think the `onAttach()` approach is the right one (with one small note inline). And I would keep the message here as a standalone, not using the helper function I suggested.

We should also think about when we want to make this defunct, and say that in the deprecation message.

**Comment by stephhazlitt** [comment]
OK, I'll change it up with {lifecycle}. When _do_ we want to make this defunct&mdash;3 months?

**Comment by ateucher** [comment]
That sounds about right to me. We can put this plus the removal of rgeos into the next release so they go together...

**Comment by stephhazlitt** [comment]
Turns out we get a lot of mileage from just updating the fxn documentation for `bc_bound_hres()` since most everything else `#' @inheritParams bc_bound_hres`. I did end up re-creating the `shortcuts.R` file, as I removed all of the `sp` examples from the docs (e.g. no more `#' my_layer_sp <- airzones(class = 'sp')`). This was inspired by the recommendation to use `#'  @keywords internal` to remove any fully deprecated function from the documentation index in the [{lifecycles} communicate article](https://lifecycle.r-lib.org/articles/communicate.html). 

Sorry for the big PR/review, turns our deprecating `sp` touches a lot of files in {bcmaps} 😧

**Review by ateucher — COMMENTED** [review]
This is looking really good!

I realized that if anyone calls a layer-getting function specifying the class, even if it's `class = "sf"` (e.g., `airzones(class = "sf")` this will fail in the future once we remove the argument. So I think we need to deprecate the argument like demonstrated [here](https://lifecycle.r-lib.org/reference/deprecated.html).

I think we can put something like this in make_shortcuts():

```r
{fn_name} <- function(class = deprecated(), ask = interactive(), force = FALSE) {{

            if (lifecycle::is_present(class)) {{
              deprecate_sp('{fn_name}(class)')
              class <- match.arg(class, choices = c('sf', 'sp'))
            }}

            get_layer('{fn_name}', class = class, ask = ask, force = force)
         }}
```

I guess we are actually deprecating two things: package-wide we are deprecating support for `Spatial` objects, and we are also deprecating the `class` argument in a bunch of functions. I think we need to communicate both without overwhelming the user with warnings... maybe just a more verbose `details` arg?

File: R/shortcuts.R [hunk 2]
```diff
@@ -1,4 +1,4 @@
-# Copyright 2022 Province of British Columbia
+# Copyright 2023 Province of British Columbia
```
Comment (boshek) [inline R/shortcuts.R:1]: What about dropping the date? More folks seem to be doing this.

File: R/zzz.R [hunk 3]
```diff
@@ -16,3 +16,21 @@ bcmaps_env <- new.env(parent = emptyenv())
   assign("bcmaps_update_message", FALSE, envir = bcmaps_env)
 }
 
+
+bcmapsStartupMessage <- function() {
+  msg <- "Support for Spatial objects (`sp`) was deprecated in {bcmaps} v1.2.0, and will be removed altogether in the Summer 2023. Please use `sf` objects with {bcmaps}."
```
Comment (boshek) [inline R/zzz.R:21]: I think we maybe try to be less precise with this estimate and just say something like "a future release"

File: R/bcmaps-package.R [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+
+
+#' @keywords internal
+"_PACKAGE"
+
+#' @import sf
+#' @importFrom utils download.file unzip
+
+## usethis namespace: start
+#' @importFrom lifecycle deprecated
```
Comment (boshek) [inline R/bcmaps-package.R:21]: are we actually using this function anywhere? Sorry if I missed it.

File: README.md [hunk 5]
```diff
@@ -154,6 +158,9 @@ ggplot() +
 
 ### Spatial (sp) objects
 
+> ***IMPORTANT NOTE** All support for Spatial objects (`sp`) is
+> deprecated in {bcmaps} v1.2.0. Please use `sf` objects with {bcmaps}.*
+
 If you aren’t using the `sf` package and prefer the old standard
 [`sp`](https://cran.r-project.org/package=sp) way of doing things, set
 `class = "sp"` in either `get_layer` or the shortcut functions:
```
Comment (boshek) [inline README.md:166]: Should we just delete this text rather than keeping it around?

File: R/zzz.R [hunk 6]
```diff
@@ -16,3 +16,21 @@ bcmaps_env <- new.env(parent = emptyenv())
...
+  # startup message
+  msg <- bcmapsStartupMessage()
+  packageStartupMessage(msg)
+  invisible()
+}
+
+deprecate_sp <- function(what){
+  lifecycle::deprecate_warn( when = "1.2.0",
+                             what = what,
+                             details = "Please use `sf` objects with {bcmaps}, support for Spatial objects (sp) will be removed in Summer 2023.")
```
Comment (boshek) [inline R/zzz.R:35]: Same as above.

File: R/make_shortcuts.R [hunk 7]
```diff
@@ -23,7 +23,7 @@
 #'
 #' @examples
 #' \dontrun{
-#' make_shortcut()
+#' make_shortcuts()
```
Comment (boshek) [inline R/make_shortcuts.R:26]: good catch

File: R/get_data.R [hunk 8]
```diff
@@ -22,9 +24,6 @@
 #' @examples
 #' \dontrun{
 #'  get_layer("bc_bound_hres")
-#'
-#'  # As a "Spatial" (sp) object
-#'  get_layer("watercourses_15M")
 #' }
 get_layer <- function(layer, class = c("sf", "sp"), ask = TRUE, force = FALSE) {
```
Comment (boshek) [inline R/get_data.R:28]: agree that it could go here and as long as the message comes across, maybe it is ok that it is at this function level?

**Review by boshek — APPROVED** [review]
Looks great! Nice work. Thanks for doing this.

File: R/shortcuts.R (hunk 2 above)
Comment (stephhazlitt) [reply R/shortcuts.R:1]: I'll have a look at the Apache template (or other new bcgov code), happy to drop if that is the new way.

File: R/zzz.R (hunk 3 above)
Comment (stephhazlitt) [reply R/zzz.R:21]: I think we want to be a little informative re: how quickly we want users to adapt, given that we are planning a reasonably short deprecation window?

File: R/bcmaps-package.R (hunk 4 above)
Comment (stephhazlitt) [reply R/bcmaps-package.R:21]: Yes, we are using it for all the deprecation pieces.

File: README.md (hunk 5 above)
Comment (stephhazlitt) [reply README.md:166]: Good idea, I'll remove it all.

File: R/zzz.R (hunk 6 above)
Comment (stephhazlitt) [reply R/zzz.R:35]: Not sure what same this comment is referring to? This is the deprecate wrapper function we are using throughout to deprecate sp?

**Comment by stephhazlitt** [comment]
Summarizing the desired approach for deprecating support for `sp`:

– `self_union()` & `get_poly_attribute()`: deprecating whole fxns
– `transform_bc_albers.Spatial` & `fix_geo_problems.Spatial`: deprecating the Spatial method
– `bc_bbox()`: deprecating the argument value `class = "sp"`
– `get_layer()` which is used for 35 shortcut functions + `combine_nr_rd()`: deprecating the `class` argument 
– 8 one off fxns that get a layer w/o using `get_layer()` 👇: deprecating the `class` argument

Screenshot 2023-04-03 at 8 01 30 PM

For get_layer()`, the best outcome for the user would be to play with the `env` and `caller_env` arguments in `deprecate_warn()` to raise the level of the warning to the calling shortcut function.

File: R/raster_by_poly.R [hunk 9]
```diff
@@ -15,6 +15,11 @@
 #' @export
 raster_by_poly <- function(raster_layer, poly, poly_field, summarize = FALSE,
                            parallel = FALSE) {
+  # I'm not deprecating this right now as both raster and sp are not
```
Comment (boshek) [inline R/raster_by_poly.R:18]: Do you think this comment might be better put into the function documentation?

**Comment by ateucher** [comment]
That's a great summary @stephhazlitt, thanks! I'm also deprecating the whole `fix_geo_problems()` function as for `sf` objects it's basically just a wrapper around `sf::st_make_valid()` - it only really had value for `Spatial` objects.

File: R/raster_by_poly.R (hunk 9 above)
Comment (ateucher) [reply R/raster_by_poly.R:18]: I don't think so... if it's not marked for deprecation then I would say it's clear enough it won't be. Just a note to ourselves when we go to make them defunct

**Comment by ateucher** [comment]
Ok - all green and I think this is ready for another review. @stephhazlitt I can't request a review from you (probably because you are the main PR author), but if you can have a look that would be awesome.

**Comment by ateucher** [comment]
And of course making a reprex found a bug. Just pushed a fix, but here's some example output:

``` r
library(bcmaps)
#> Loading required package: sf
#> Linking to GEOS 3.11.0, GDAL 3.5.3, PROJ 9.1.0; sf_use_s2() is TRUE
#> Support for Spatial objects (`sp`) was deprecated in {bcmaps} v1.2.0, and will be removed altogether in the Summer 2023. Please use `sf` objects with {bcmaps}.

# ensure it always warns, for demo purposes
withr::local_options(lifecycle_verbosity = "warning")

az <- airzones()
#> airzones was updated on NULL
az <- airzones("sp")
#> Warning: The `class` argument of `airzones()` is deprecated as of bcmaps 1.2.0.
#> ℹ Please use `sf` objects with {bcmaps}, support for Spatial objects (sp) will
#>   be removed in Summer 2023.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.

x <- get_layer("bc_cities")
x <- get_layer("bc_cities", class = "sp")
#> Warning: The `class` argument of `get_layer()` is deprecated as of bcmaps 1.2.0.
#> ℹ Please use `sf` objects with {bcmaps}, support for Spatial objects (sp) will
#>   be removed in Summer 2023.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.
x <- get_layer("bc_cities", class = "sf")
#> Warning: The `class` argument of `get_layer()` is deprecated as of bcmaps 1.2.0.
#> ℹ Please use `sf` objects with {bcmaps}, support for Spatial objects (sp) will
#>   be removed in Summer 2023.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.

nr_rc <- combine_nr_rd(class = "sp")
#> Warning: The `class` argument of `combine_nr_rd()` is deprecated as of bcmaps 1.2.0.
#> ℹ Please use `sf` objects with {bcmaps}, support for Spatial objects (sp) will
#>   be removed in Summer 2023.
#> Call `lifecycle::last_lifecycle_warnings()` to see where this warning was
#> generated.
```

<sup>Created on 2023-04-06 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by stephhazlitt** [comment]
This is SO close @ateucher, thanks for the changes. A couple of comments:

- I think we have a bit of inconsistency re: versioning plans. The start up message gives the deprecation timeline as `Support for Spatial objects (sp) was deprecated in {bcmaps} v1.2.0, and will be removed altogether in the Summer 2023` and the docs say `class = sp is no longer supported and as of bcmaps 1.2.0 (planned for Summer 2023) the class argument will be removed`. I think I assumed we would do a release with the deprecation changes (which would bump from 1.1 -> 1.2), and then a subsequent release in the summer with the full removal? Either way, we should align the messages.
- `#' @keywords internal` IIUC this removes the deprecated function from the help index (and pkgdown site if there is one), removing the user user-facing documentation. Users can only access the deprecation note if they are already users of the fxn and type `?deprecate_fxn` in the console. I wonder if in our use case, this is a fast/abrupt design and it we should leave the deprecated fxns in the index?
- This could just be a temporary local issue, but `browseVignettes("bcmaps")` is not working for me on the PR (but it does with the main branch).

**Comment by ateucher** [comment]
Thanks @stephhazlitt. 

> * I think we have a bit of inconsistency re: versioning plans. The start up message gives the deprecation timeline as `Support for Spatial objects (sp) was deprecated in {bcmaps} v1.2.0, and will be removed altogether in the Summer 2023` and the docs say `class = sp is no longer supported and as of bcmaps 1.2.0 (planned for Summer 2023) the class argument will be removed`. I think I assumed we would do a release with the deprecation changes (which would bump from 1.1 -> 1.2), and then a subsequent release in the summer with the full removal? Either way, we should align the messages.

Great catch re the messaging - You're right, it's deprecated in 1.2.0 and will be removed in the summer. I'll align the messages.

> * `#' @keywords internal` IIUC this removes the deprecated function from the help index (and pkgdown site if there is one), removing the user user-facing documentation. Users can only access the deprecation note if they are already users of the fxn and type `?deprecate_fxn` in the console. I wonder if in our use case, this is a fast/abrupt design and it we should leave the deprecated fxns in the index?

I think I'm ok removing the function from the index... I don't think we want anyone discovering it now and starting to use it if it's going to be removed in a few months.

> * This could just be a temporary local issue, but `browseVignettes("bcmaps")` is not working for me on the PR (but it does with the main branch).

I'll double-check the vignettes.

**Comment by ateucher** [comment]
Ok @stephhazlitt I think I've addressed your comments

**Comment by stephhazlitt** [comment]
Mmmmmmmerge 🚀

File: README.Rmd [hunk 10]
```diff
@@ -36,15 +36,19 @@ An [R](https://www.r-project.org) package of spatial map layers for British Colu
...
-administrative boundaries, natural resource management boundaries, watercourses etc. 
-All layers are available in the
-[BC Albers](https://spatialreference.org/ref/epsg/nad83-bc-albers/) projection, which is the B.C. Government standard as `sf` or `Spatial` objects. 
+administrative boundaries, natural resource management boundaries, watercourses, census boundaries, etc. All layers are available as `sf` objects in the
+[BC Albers](https://spatialreference.org/ref/epsg/nad83-bc-albers/) projection, which is the B.C. Government standard. 
 
 Most layers are assessed directly from the [B.C. Data Catalogue](https://catalogue.data.gov.bc.ca/) using the [bcdata](https://github.com/bcgov/bcdata) R package. See each layers individual help file for more detail.
 
+
+> _**IMPORTANT NOTE** Support for Spatial objects (`sp`) is deprecated in {bcmaps} v1.2.0, and will be removed in Summer 2023. Please use `sf` objects with {bcmaps}._
```
Comment (boshek) [inline README.Rmd:45]:
```suggestion
> _**IMPORTANT NOTE** Support for Spatial objects (`sp`) is deprecated in {bcmaps} v1.2.0, and will be removed in Summer 2023. Please use `sf` objects with {bcmaps}._ A discussion on the evolution of the spatial software stack in R can be found here: https://r-spatial.org/r/2022/04/12/evolution.html. 
```

**Review by boshek — APPROVED** [review]
LGTM. One suggestion.

File: README.Rmd (hunk 10 above)
Comment (stephhazlitt) [reply README.Rmd:45]: Great suggestion.

**Pull request merged**